### PR TITLE
Feature: Action to rearrange all components

### LINKED ISSUE #355: Action to rearrange all components
## Description

**As a** a`User`,

**I want** `to rearrange all components positions`

**so that** `I can have a readable diagram`.

Create an action bar with one button `Rearrange` at the bottom right of the page.

:warning: If user open the component details panel, make the action bar moving on the left of its panel, to let it visible.

## Acceptance Criteria

 **Scenario** | `[the name for the behavior that will be described]`
 ------------ | -------------------------------------------------
 **Given**    | `I go to my diagram page`
 **When**     | `I click on rearrange button`
 **Then**     | `I expect all my components are moving and its rearranged`

## Resources:

Icon to use: `sitemap`

Example of action bar: 
image

### PR DESCRIPTION
**Pull request by thibaudlabat**
Users can now *automatically* rearrange all components positions so that they can have a readable diagram.

image

Closes  #355

### CODE REVIEW

File: src/layouts/ModelizerDrawLayout.vue [hunk 1]
```diff
@@ -92,6 +112,19 @@ async function initView() {
...
 });
+
+onUnmounted(() => {
+  pluginInitSubscription.unsubscribe();
+});
+
+/**
+ * Rearrange components on click.
+ */
+async function arrangeComponentsPosition() {
```
Comment (Zorin95670) [inline src/layouts/ModelizerDrawLayout.vue:123]: put this function before `onMounted`

File: src/layouts/ModelizerDrawLayout.vue [hunk 2]
```diff
@@ -92,6 +112,19 @@ async function initView() {
...
+  await data.plugin.arrangeComponentsPosition();
+  data.plugin.draw(drawTarget);
+}
+
```
Comment (Zorin95670) [inline src/layouts/ModelizerDrawLayout.vue:127]: Missing `changelog.md` update in another commit and dependencies update in another commit too.

File: src/layouts/ModelizerDrawLayout.vue [hunk 3]
```diff
@@ -92,6 +112,19 @@ async function initView() {
 onMounted(() => {
   initView();
 });
+
+onUnmounted(() => {
```
Comment (Zorin95670) [inline src/layouts/ModelizerDrawLayout.vue:116]: Why do you `unsubscribe` if no `subscribe` is done ?
I think `pluginInitSubscription` is useless.

File: src/layouts/ModelizerDrawLayout.vue (hunk 1 above)
Comment (Zorin95670) [inline src/layouts/ModelizerDrawLayout.vue:123]: missing unit test.

Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:123]: done

File: src/layouts/ModelizerDrawLayout.vue (hunk 3 above)
Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:116]: I don't remember having written this code. Must have been placed here by mistake.

I removed it.

File: src/layouts/ModelizerDrawLayout.vue (hunk 1 above)
Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:113]: done

File: src/layouts/ModelizerDrawLayout.vue (hunk 2 above)
Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:118]: changelog: done
dependencies: done

Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:118]: e2e tests: done

File: tests/unit/layouts/ModelizerDrawLayout.spec.js [hunk 4]
```diff
@@ -142,5 +142,25 @@ describe('Test page component: ModelizerDrawLayout', () => {
 
       expect(Notify.create).not.toHaveBeenCalled();
     });
+
+    it('should arrange a component position and redraw on click', async () => {
```
Comment (Zorin95670) [inline tests/unit/layouts/ModelizerDrawLayout.spec.js:146]: Its in wrong describe, please put it in a new `describe('Test function: arrangeComponentsPosition', () => {})`
outside the `describe('Test function: initView', () => {`

File: tests/unit/layouts/ModelizerDrawLayout.spec.js [hunk 5]
```diff
@@ -142,5 +142,25 @@ describe('Test page component: ModelizerDrawLayout', () => {
...
+      let awaitedArrangeComponentsPosition = false;
+      wrapper.vm.data.plugin = {
```
Comment (Zorin95670) [inline tests/unit/layouts/ModelizerDrawLayout.spec.js:148]: Set mock before when we : `jest.mock('src/composables/PluginManager', () => ({`
just like :
```js
parse: jest.fn(),
draw: jest.fn(),
```

File: cypress/support/step_definitions/html.js [hunk 6]
```diff
@@ -142,8 +142,9 @@ Then('I expect {string} height is {int}', (templateSelector, height) => {
 Then('I expect {string} to be at position {int},{int}', (templateSelector, x, y) => {
```
Comment (Zorin95670) [inline cypress/support/step_definitions/html.js:142]: Rule is already use to check on screen diagrams position at diagrams page.
Prefere to have an another rules like 
`I expect component {string} to be at position {int},{int}`

File: changelog.md [hunk 7]
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Draw page:
   * Add variables panel in left drawer.
   * Rename component id.
+  * Add a button to reposition all components automatically
```
Comment (Zorin95670) [inline changelog.md:28]: .

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 4 above)
Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:146]: done

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 5 above)
Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:148]: Then how could I check that plugin.arrangeComponentsPosition() has been properly awaited ?
Or maybe I don't need to check this ?

File: changelog.md (hunk 7 above)
Comment (thibaudlabat) [reply changelog.md:28]: done

File: cypress/support/step_definitions/html.js (hunk 6 above)
Comment (thibaudlabat) [reply cypress/support/step_definitions/html.js]: done

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 5 above)
Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js]: done

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 4 above)
Comment (Zorin95670) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js]: not done, bad correctione see with you by message

Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:146]: done

**Review by Zorin95670 — DISMISSED** [review]

File: src/layouts/ModelizerDrawLayout.vue [hunk 8]
```diff
@@ -18,6 +18,21 @@
     </q-page-container>
     <git-authentication-dialog :project-name="projectName" />
     <git-add-remote-dialog :project-name="projectName" />
+
+    <q-page-sticky :offset="[20, 20]">
```
Comment (k-y0u) [inline src/layouts/ModelizerDrawLayout.vue:22]: should be in `ModelizerDrawPage.vue > q-page` as done in `DiagramsPage`.

File: src/layouts/ModelizerDrawLayout.vue [hunk 9]
```diff
@@ -89,6 +104,16 @@ async function initView() {
   ]);
 }
 
+/**
+ * Rearrange components on click,
```
Comment (k-y0u) [inline src/layouts/ModelizerDrawLayout.vue:108]: delete `on click`

File: src/layouts/ModelizerDrawLayout.vue [hunk 10]
```diff
@@ -89,6 +104,16 @@ async function initView() {
...
+ * and then redraw the whole view.
+ * @param {string} targetId - Id of the drawing target.
+ */
+async function arrangeComponentsPosition(targetId) {
```
Comment (k-y0u) [inline src/layouts/ModelizerDrawLayout.vue:112]: parameter not necessary (cf. `ModelizerDrawPage.vue`)

File: tests/unit/layouts/ModelizerDrawLayout.spec.js [hunk 11]
```diff
@@ -134,13 +134,25 @@ describe('Test page component: ModelizerDrawLayout', () => {
...
-      Notify.create = jest.fn();
+      await wrapper.vm.initView();
 
+      expect(notifySpy).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('Test function: arrangeComponentsPosition', () => {
+    it('should arrange a component position and redraw on click', async () => {
+      // We need this to load the plugin.
```
Comment (k-y0u) [inline tests/unit/layouts/ModelizerDrawLayout.spec.js:147]: `initView()` is already done in `onMounted`, so not necessary here

File: changelog.md [hunk 12]
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Draw page:
   * Add variables panel in left drawer.
   * Rename component id.
+  * Add a button to reposition all components automatically.
```
Comment (k-y0u) [inline changelog.md:28]: reposition -> rearrange

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: tests/unit/layouts/ModelizerDrawLayout.spec.js [hunk 13]
```diff
@@ -134,13 +134,25 @@ describe('Test page component: ModelizerDrawLayout', () => {
     });
 
     it('should do nothing when there is no plugin', async () => {
-      PluginManager.getPluginByName = jest.fn(() => null);
+      const notifySpy = jest.spyOn(Notify, 'create');
```
Comment (k-y0u) [inline tests/unit/layouts/ModelizerDrawLayout.spec.js:137]: Is this change bring something ? If not, no need to update. better stay with the usual way to do.

File: cypress/support/step_definitions/url.js [hunk 14]
```diff
@@ -14,7 +14,7 @@ Given('I reload to {string}', (urlTemplate) => {
 });
 
 Given('I wait until the application is loaded', () => {
-  cy.url({ timeout: 15000 }).should('not.equal', '/squash');
+  cy.url({ timeout: 30000 }).should('not.match', /\/splash/);
```
Comment (k-y0u) [inline cypress/support/step_definitions/url.js:17]: isn't it too much ? Why so much ?

File: src/layouts/ModelizerDrawLayout.vue (hunk 10 above)
Comment (thibaudlabat) [reply src/layouts/ModelizerDrawLayout.vue:112]: removed

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 11 above)
Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:147]: deleted

File: changelog.md (hunk 12 above)
Comment (thibaudlabat) [reply changelog.md:28]: changed

File: cypress/support/step_definitions/url.js (hunk 14 above)
Comment (thibaudlabat) [reply cypress/support/step_definitions/url.js:17]: No specific reason, I roll it back.

File: tests/unit/layouts/ModelizerDrawLayout.spec.js (hunk 13 above)
Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:137]: `PluginManager.getPluginByName = jest.fn(() => null);` contaminates a global object and that is not cleaned up when the test finishes.

Removing the line solved the problem and the test still passes. Reading the code again, I am wondering what really was the purpose of this line if the test passes without it?

Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js:137]: Anyway I will roll it back as I am moving my code from `ModelizerDrawLayout.vue` to `ModelizerDrawPage.vue`, and also moving the tests accordingly.

Comment (thibaudlabat) [reply tests/unit/layouts/ModelizerDrawLayout.spec.js]: rolled back

**Comment by thibaudlabat** [comment]
Can't update `cypress  =12.17.4  →  =13.0.0` because of incompatible packages.

**Review by justinehell — APPROVED** [review]

**Pull request merged**